server/apple_notification_server.go

```diff
@@ -46,94 +46,98 @@ func (me *AppleNotificationServer) Initialize() bool {
+		data.MutableContent()
 
 		if len(msg.ChannelName) > 0 && msg.Version == "v2" {
-			payload.AlertTitle(msg.ChannelName)
-			payload.AlertBody(emoji.Sprint(msg.Message))
-			payload.Custom("channel_name", msg.ChannelName)
+			data.AlertTitle(msg.ChannelName)
+			data.AlertBody(emoji.Sprint(msg.Message))
+			data.Custom("channel_name", msg.ChannelName)
 		} else {
-			payload.Alert(emoji.Sprint(msg.Message))
+			data.Alert(emoji.Sprint(msg.Message))
 
 			if len(msg.ChannelName) > 0 {
-				payload.Custom("channel_name", msg.ChannelName)
+				data.Custom("channel_name", msg.ChannelName)
 			}
 		}
 	} else {
-		payload.ContentAvailable()
+		data.ContentAvailable()
 	}
 
-	payload.Custom("type", msg.Type)
+	incrementNotificationTotal(PUSH_NOTIFY_APPLE, pushType)
+	data.Custom("type", pushType)
 
 	if len(msg.ChannelId) > 0 {
-		payload.Custom("channel_id", msg.ChannelId)
-		payload.ThreadID(msg.ChannelId)
+		data.Custom("channel_id", msg.ChannelId)
+		data.ThreadID(msg.ChannelId)
 	}
 
 	if len(msg.TeamId) > 0 {
-		payload.Custom("team_id", msg.TeamId)
+		data.Custom("team_id", msg.TeamId)
 	}
 
 	if len(msg.SenderId) > 0 {
-		payload.Custom("sender_id", msg.SenderId)
+		data.Custom("sender_id", msg.SenderId)
 	}
 
 	if len(msg.PostId) > 0 {
-		payload.Custom("post_id", msg.PostId)
+		data.Custom("post_id", msg.PostId)
 	}
 
 	if len(msg.RootId) > 0 {
-		payload.Custom("root_id", msg.RootId)
+		data.Custom("root_id", msg.RootId)
 	}
 
 	if len(msg.OverrideUsername) > 0 {
-		payload.Custom("override_username", msg.OverrideUsername)
+		data.Custom("override_username", msg.OverrideUsername)
 	}
 
 	if len(msg.OverrideIconUrl) > 0 {
-		payload.Custom("override_icon_url", msg.OverrideIconUrl)
+		data.Custom("override_icon_url", msg.OverrideIconUrl)
 	}
 
 	if len(msg.FromWebhook) > 0 {
-		payload.Custom("from_webhook", msg.FromWebhook)
+		data.Custom("from_webhook", msg.FromWebhook)
 	}
 
 	if me.AppleClient != nil {
 		LogInfo(fmt.Sprintf("Sending apple push notification for device=%v and type=%v", me.ApplePushSettings.Type, msg.Type))
 		start := time.Now()
 		res, err := me.AppleClient.Push(notification)
-		observeAPNSResponse(time.Since(start).Seconds())
+		observerNotificationResponse(PUSH_NOTIFY_APPLE, time.Since(start).Seconds())
 		if err != nil {
 			LogError(fmt.Sprintf("Failed to send apple push sid=%v did=%v err=%v type=%v", msg.ServerId, msg.DeviceId, err, me.ApplePushSettings.Type))
-			incrementFailure(me.ApplePushSettings.Type)
+			incrementFailure(PUSH_NOTIFY_APPLE, pushType, "RequestError")
 			return NewErrorPushResponse("unknown transport error")
 		}
 
 		if !res.Sent() {
-			if res.Reason == "BadDeviceToken" || res.Reason == "Unregistered" || res.Reason == "MissingDeviceToken" || res.Reason == "DeviceTokenNotForTopic" {
+			if res.Reason == apns.ReasonBadDeviceToken || res.Reason == apns.ReasonUnregistered || res.Reason == apns.ReasonMissingDeviceToken || res.Reason == apns.ReasonDeviceTokenNotForTopic {
```

Comment:
Not sure I understand this. Couldn't we just check the `msg.Type` here? Why do we increment removal?

Reply:
`msg.Type` is to establish if the notification that we are sending is to clear previous sent notifications or if is a notification with a message.

On the other hand we do increment removal as it was a previous metric that is actually used by shift and I wanted to keep it just in case, the reality is that is a failure and all that we need in terms of functionality is the removal response to avoid sending further notifications to an invalid device. I'm 0/5 on removing the removal metric.

Reply:
I would say remove it as it appears to not make much sense. It's not tracking removals, it's tracking a subset of failures that have nothing to do with removals.

Reply:
what do you mean? it is tracking removals, every single failure that matches the ones there is considered a removed device

Reply:
`res.Reason == apns.ReasonBadDeviceToken` that only happens on removals? Maybe I just don't understand the terminology here. I thought removal was PUSH_TYPE_CLEAR?

Reply:
`res.Reason == apns.ReasonBadDeviceToken` means that the push notification service is returning that the device token is not valid so we need to remove it to avoid sending future notifications.

`PUSH_TYPE_CLEAR`: means that it is a notification that is sent to tell the device to clear previous message notifications received.

Reply:
Alright, then this is just me misunderstanding. Thanks for clarifying.